matsim/src/main/java/org/matsim/core/population/io/ParallelPopulationReaderMatsimV6.java

```diff
@@ -0,0 +1,222 @@
+ * *********************************************************************** *
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *   See also COPYING, LICENSE and WARRANTY file                           *
+ *                                                                         *
+ * *********************************************************************** */
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.matsim.api.core.v01.Id;
+import org.matsim.api.core.v01.Scenario;
+import org.matsim.api.core.v01.population.Person;
+import org.xml.sax.Attributes;
+import org.xml.sax.helpers.AttributesImpl;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Stack;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+
+/**
+ * Parallel implementation of the PopulationReaderMatsimV6. The main thread only reads
+ * the file and creates empty person objects which are added to the population to ensure
+ * that their order is not changed. Note that this approach is not compatible with
+ * population streaming. When this feature is activated, the non-parallel reader is used.
+ * The parallel threads interpret the xml data for each person.
+ *
+ * @author cdobler, steffenaxer
+ */
+/* deliberately package */  class ParallelPopulationReaderMatsimV6 extends PopulationReaderMatsimV6 {
+
+	static final Logger log = LogManager.getLogger(ParallelPopulationReaderMatsimV6.class);
+
+	private final boolean isPopulationStreaming;
+	private final int numThreads;
+	private final BlockingQueue<List<Tag>> queue;
+	private Thread[] threads;
+	private List<Tag> currentPersonXmlData;
+
+	private final String inputCRS;
+	private final String targetCRS;
+
+	private boolean reachedPersons = false;
+
+
+	public ParallelPopulationReaderMatsimV6(
+			final String inputCRS,
+			final String targetCRS,
+			final Scenario scenario) {
+		super(inputCRS, targetCRS, scenario);
+		this.inputCRS = inputCRS;
+		this.targetCRS = targetCRS;
+
+		/*
+		 * Check whether population streaming is activated
+		 */
+
+		if (scenario.getPopulation() instanceof StreamingPopulationReader.StreamingPopulation) {
+			log.warn("Population streaming is activated - cannot use " + ParallelPopulationReaderMatsimV6.class.getName() + "!");
+
+			this.isPopulationStreaming = true;
+			this.numThreads = 1;
+			this.queue = null;
+
+		} else {
+			isPopulationStreaming = false;
+
+			if (scenario.getConfig().global().getNumberOfThreads() > 0) {
+				this.numThreads = scenario.getConfig().global().getNumberOfThreads();
```

Comment:
it might be the best option to use `global().getNumberOfThreads()` as there is no other, more specific setting. But I wonder if 40 threads really bring a difference to — say — at most 4 or 8. Maybe hard-code a limit (`Math.min(4, config.global().getNumberOfThreads())`

Reply:
Limited to 4